pkg/controllers/consolidation/controller.go

```diff
@@ -0,0 +1,665 @@
+			if c.cluster.LastNodeDeletionTime().Before(stabilizationTime) {
+				if c.ProcessCluster(ctx) == ProcessResultNothingToDo {
+					c.lastConsolidationState = c.cluster.ClusterConsolidationState()
+				}
+			}
+		}
+	}
+}
+
+// candidateNode is a node that we are considering for consolidation along with extra information to be used in
+// making that determination
+type candidateNode struct {
+	*v1.Node
+	instanceType   cloudprovider.InstanceType
+	provisioner    *v1alpha5.Provisioner
+	disruptionCost float64
+	pods           []*v1.Pod
+}
+
+type ProcessResult byte
+
+const (
+	ProcessResultNothingToDo ProcessResult = iota
+	ProcessResultFailed
+	ProcessResultConsolidated
+)
+
+// ProcessCluster is exposed for unit testing purposes
+func (c *Controller) ProcessCluster(ctx context.Context) ProcessResult {
+	candidates := c.candidateNodes(ctx)
+	if len(candidates) == 0 {
+		return ProcessResultNothingToDo
+	}
+
+	emptyNodes := lo.Filter(candidates, func(n candidateNode, _ int) bool { return len(n.pods) == 0 })
+	// first see if there are empty nodes that we can delete immediately, and if so delete them all at once
+	if len(emptyNodes) > 0 {
+		c.performConsolidation(ctx, consolidationAction{
+			oldNodes: lo.Map(emptyNodes, func(n candidateNode, _ int) *v1.Node { return n.Node }),
+			result:   consolidateResultDeleteEmpty,
+		})
+		return ProcessResultConsolidated
+	}
+
+	pdbs, err := NewPDBLimits(ctx, c.kubeClient)
+	if err != nil {
+		logging.FromContext(ctx).Errorf("Tracking PodDisruptionBudgets, %s", err)
+		return ProcessResultFailed
+	}
+
+	// the remaining nodes are all non-empty, so we just consolidate the first one that we can
+	sort.Slice(candidates, byNodeDisruptionCost(candidates))
+	for _, node := range candidates {
+		// is this a node that we can terminate?  This check is meant to be fast so we can save the expense of simulated
+		// scheduling unless its really needed
+		if err = c.canBeTerminated(node, pdbs); err != nil {
+			continue
+		}
+		action := c.nodeConsolidationActions(ctx, node)
+		if action.result == consolidateResultDelete || action.result == consolidateResultShrink {
+			// perform the first consolidation we can since we are looking at nodes in ascending order of disruption cost
+			c.performConsolidation(ctx, action)
+			return ProcessResultConsolidated
+		}
+	}
+	return ProcessResultNothingToDo
+}
+
+// candidateNodes returns nodes that appear to be currently consolidatable based off of their provisioner
+func (c *Controller) candidateNodes(ctx context.Context) []candidateNode {
+	var nodes []candidateNode
+	uninitializedNodeExists := false
+	c.cluster.ForEachNode(func(n *state.Node) bool {
+		// skip any nodes that we can't determine the instance of or for which we don't have consolidation enabled
+		if n.InstanceType == nil || n.Provisioner == nil || n.Provisioner.Spec.Consolidation == nil || !aws.BoolValue(n.Provisioner.Spec.Consolidation.Enabled) {
```

Comment:
What do you think about using a label on the node instead of looking up the provisioner? We used to do a similar approach for emptiness/expiry. This helps to decouple consolidation from provisioners and gives users a hook to elect (or exempt) a node from consolidation. 

We will have the issue that a provisioner that toggles consolidation may drift from nodes, but we have other work in flight to catch this case.

Reply:
That's a large departure from the design document.  Someone can experiment with other options in a follow-on PR.

Reply:
https://github.com/aws/karpenter/pull/2123#issuecomment-1188258308